### ISSUE TITLE
[Detector Support]: Coral Detection Stuck, Killed, Doesn't Come Back

### ISSUE THREAD

**Issue by Sammy1Am** [blakeblackshear/frigate#7323]
### Describe the problem you are having

Had Frigate up and running on an Unraid system with an M.2 PCIe Coral TPU working fine; weeks at a time with no restarts.  Recently moved from Unraid to Ubuntu Server 23, and installed another M.2 NVMe drive in my server.

Since the move, Frigate seems to start up and run fine for a little while, but after several hours (like 4-12ish) `frigate.watchdog` says that detection appears to be stuck (see logs), attempts to stop it, fails, force kills it (maybe?) and then Frigate can no longer detect the TPU.  This causes Frigate to exit, the docker container restarts, but when it restarts it's unable to find the TPU and immediately fails again.

Each time Frigate restarts, the kernel log shows:
```
apex 0000:01:00.0: Error in device open cb: -110
apex 0000:01:00.0: RAM did not enable within timeout (12000 ms)
```

Restarting the server resolves the problem for another several hours.

There are a couple power-saving options that potentially can lead to the "RAM did not enable" error, but I attempted fixing those without success.  As far I can can tell there aren't any other system events happening when the detection appears to get stuck, so it seems to be an issue within the Frigate/Coral area.

- Has something changed in a recent version of Frigate that would make rolling back worth a try (and how far back can I go with my current database, config, etc.)?
- Is force-killing the detection process somehow rendering the TPU unavailable, or does the TPU become unavailable first and force-killing the process just doesn't fix it?
- This is maybe more a Coral AI question, but is there some way to manually reset/restart the TPU driver or device without restarting the whole server?  This isn't really a solution per se, but might help indicate what's breaking in the first place.

### Version

0.12.1-367D724

### Frigate config file

```yaml
detectors:
  coral1:
    type: edgetpu
    device: pci:0
```

### docker-compose file or Docker CLI command

```yaml
Frigate:
    restart: unless-stopped
    image: ghcr.io/blakeblackshear/frigate:stable
    shm_size: "512mb" # update for your cameras based on calculation above
    devices:
      - /dev/apex_0:/dev/apex_0 # passes a PCIe Coral, follow driver instructions here https://coral.ai/docs/m2/get-started/#2a-on-linux
      - /dev/dri/renderD128 # for intel hwaccel, needs to be updated for your hardware
    volumes:
      - /etc/localtime:/etc/localtime:ro
      - /mnt/appdata/frigate/config.yml:/config/config.yml:ro
      - /mnt/appdata/frigate/nginx.conf:/usr/local/nginx/conf/nginx.conf:ro
      - /mnt/appdata/frigate:/db
      - /mnt/storage/security/frigate:/media/frigate
      - type: tmpfs # Optional: 1GB of memory, reduces SSD/SD Card wear
        target: /tmp/cache
        tmpfs:
          size: 1000000000
    networks:
      tf0:
        aliases:
          - frigate.docc
```

### Relevant log output

```shell
{"log":"2023-07-28 18:00:15.801793561  ::ffff:172.22.0.135 - - [28/Jul/2023:18:00:15 -0700] \"GET /api/events/1690592411.799732-8g0wm6/snapshot.jpg HTTP/1.1\" 200 119330 \"-\" \"python-requests/2.31.0\" \"-\"\n","stream":"stdout","time":"2023-07-29T01:00:15.801877756Z"}
{"log":"2023-07-28 18:19:38.490017284  [2023-07-28 18:19:38] frigate.watchdog               INFO    : Detection appears to be stuck. Restarting detection process...\n","stream":"stdout","time":"2023-07-29T01:19:38.490309693Z"}
{"log":"2023-07-28 18:19:38.490288843  [2023-07-28 18:19:38] root                           INFO    : Waiting for detection process to exit gracefully...\n","stream":"stdout","time":"2023-07-29T01:19:38.490425141Z"}
{"log":"2023-07-28 18:20:08.517789509  [2023-07-28 18:20:08] root                           INFO    : Detection process didnt exit. Force killing...\n","stream":"stdout","time":"2023-07-29T01:20:08.518009193Z"}
{"log":"2023-07-28 18:20:21.429011839  [2023-07-28 18:20:21] root                           INFO    : Detection process has exited...\n","stream":"stdout","time":"2023-07-29T01:20:21.429250761Z"}
{"log":"2023-07-28 18:20:21.456607959  [2023-07-28 18:20:21] detector.coral1                INFO    : Starting detection process: 8809\n","stream":"stdout","time":"2023-07-29T01:20:21.456723214Z"}
{"log":"2023-07-28 18:20:34.354616936  [2023-07-28 18:20:33] frigate.record                 WARNING : Unable to keep up with recording segments in cache for backyard. Keeping the 5 most recent segments out of 6 and discarding the rest...\n","stream":"stdout","time":"2023-07-29T01:20:34.354845645Z"}
{"log":"2023-07-28 18:20:34.355008379  [2023-07-28 18:20:21] frigate.detectors.plugins.edgetpu_tfl INFO    : Attempting to load TPU as pci:0\n","stream":"stdout","time":"2023-07-29T01:20:34.355168739Z"}
{"log":"2023-07-28 18:20:34.355399193  Process detector:coral1:\n","stream":"stdout","time":"2023-07-29T01:20:34.355541953Z"}
{"log":"2023-07-28 18:20:34.355410670  [2023-07-28 18:20:34] frigate.detectors.plugins.edgetpu_tfl ERROR   : No EdgeTPU was detected. If you do not have a Coral device yet, you must configure CPU detectors.\n","stream":"stdout","time":"2023-07-29T01:20:34.355574268Z"}




apex 0000:01:00.0: Error in device open cb: -110
apex 0000:01:00.0: RAM did not enable within timeout (12000 ms)
```

### Operating system

Debian

### Install method

Docker Compose

### Coral version

M.2

**Comment by skrashevich** [comment]
it's a problem with coral driver
show `lspci -nn | grep 089a` and check udev rules(most likely in file `/etc/udev/rules.d/65-apex.rules`). Make sure they're not doubled

**Comment by Sammy1Am** [comment]
Just to be clear, the Coral driver is definitely loading and functioning correctly initially.  Frigate successfully detects objects for several hours before it gets stuck.

`lspci -nn | grep 089a`:
`01:00.0 System peripheral [0880]: Global Unichip Corp. Coral Edge TPU [1ac1:089a]`

I didn't find any apex.rules, where else should I be looking?
`sudo cat /usr/lib/udev/rules.d/60-gasket-dkms.rules`:
`SUBSYSTEM=="apex", MODE="0660", GROUP="apex"`

`dmesg` only contains:
```
[   14.844503] apex 0000:01:00.0: enabling device (0000 -> 0002)
[   20.054047] apex 0000:01:00.0: Apex performance not throttled due to temperature
```

**Comment by NickM-27** [comment]
There are a lot of different posts about this issue on the Google coral repo which is a good first place to look. 

https://github.com/google-coral/edgetpu/issues/491

https://github.com/blakeblackshear/frigate/issues/5910

**Comment by Sammy1Am** [comment]
> There are a lot of different posts about this issue on the Google coral repo which is a good first place to look.

I definitely started there (but thank you for a direct link 🙂), and the issue you linked contains the two potential power-saving settings that I mentioned I had tried to no avail.  Most of the commenters (or at least OP specifically) seem to be having trouble with the device initially loading (e.g. they're seeing errors in `dmesg`); however I'm not having any trouble with the driver initially loading, and Frigate sees the TPU and works just fine for several hours before running into any issues.

`RAM did not enable within timeout` seems to be a fairly common error state for the Coral TPU unfortunately.

**Comment by NickM-27** [comment]
Yeah, I run a dual edge TPU in Unraid and have never had this issue. Do you have another slot that can be tried, tried swapping between the existing m.2?

I have really only seen these types of issues with usb coral

To answer your questions in the OP

> Has something changed in a recent version of Frigate that would make rolling back worth a try (and how far back can I go with my current database, config, etc.)?

Nothing has changed in this regard for many versions.

> Is force-killing the detection process somehow rendering the TPU unavailable, or does the TPU become unavailable first and force-killing the process just doesn't fix it?

Frigate does not try to restart or kill the process until it does not receive a response form the coral. Sometimes corals can become stuck (especially USB corals) and restarting the process can get it back in gear. If restarting doesn't work then the coral is not responding. Dmesg confirms that this is an issue occurring on the host with the coral.

> This is maybe more a Coral AI question, but is there some way to manually reset/restart the TPU driver or device without restarting the whole server? This isn't really a solution per se, but might help indicate what's breaking in the first place.

For the USB it would be just unplugging but that is not the same for m.2

**Comment by skrashevich** [comment]
Create a new file at `/etc/modprobe.d/blacklist-apex.conf` and add these two lines:
```
blacklist gasket
blacklist apex
```
then reboot system

**Comment by Sammy1Am** [comment]
@NickM-27 Thanks for the replies to my initial questions, that rules out a few things at least.

I did figure out that I can get things working again with:
```bash
echo 1 | sudo tee /sys/bus/pci/devices/[deviceid]/remove
sleep 1
echo 1 | sudo tee /sys/bus/pci/rescan
```

So it seems like the TPU is getting into some sort of unrecoverable state and resetting/power cycling it fixes it.

@skrashevich I'm not sure I'm understanding your suggestion.  The drivers seem to be loading fine and everything is working for hours at a time.  Are you suggesting I may have two different drivers present?  If that's the case, is there any way to see which drivers I have / from where to make sure I'm using the correct one?

**Comment by skrashevich** [comment]
> @skrashevich I'm not sure I'm understanding your suggestion.

Oh, the last step in previous message is lost :(
after reboot, reinstall kernel module (`apt-get install --reinstall gasket-dkms`), remove `blacklist-apex.conf` and reboot again

**Comment by Sammy1Am** [comment]
An update: installed a second TPU in the system (the mini-PCIe version on an adapter card) and things seem fine now.  I can't easily tell which device is which (unless someone knows a handy command for that), so it's not immediately clear if installing the second TPU fixed the first one, or if the second one took `apex_0` and the first one is now `apex_1`.  But this definitely seems to suggest that unless there's some difference between installing it in the M.2 slot vs PCIe, that it's:

- Not aspm. (I turned it back on after installing the second card and things are still fine)
- Not drivers. (Didn't make any changes there and things are working fine)
- Not Frigate. (I didn't change anything in the Frigate configuration since the working Unraid instance)

So, I'll close this issue, but for posterity in case anyone else comes across this:

It's heavily pointing to a hardware issue, but a weird one.  I know the card itself was good right before installing Ubuntu, and it's unlikely that installing a new OS actually ruined the card (especially if it didn't ruin the second one I put in there).  The only other hardware change I made was installing an M.2 NVMe drive in the other M.2 slot.  So my best guess is it's a power usage thing (like both devices cause enough of a dip to do something weird to the TPU), or some other communication thing (like if they're sharing a lane or something?  I don't know).

Anyway, thanks for the help guys, I've got it worked around for now, but seems like not Frigate's fault in anycase.
